### ISSUE TITLE
speculative: intended per-round call order for a full MtpSpeculative generation loop?

### ISSUE THREAD

**Issue by leon-mallett** [utilityai/llama-cpp-rs#1105]
Thanks for the new `speculative` module in `llama-cpp-2` 0.1.154 — being able to reach same-model MTP
speculative decoding from Rust is great. I'm trying to build a full generation loop with
`MtpSpeculative` on a **Qwen3.5** GGUF (the "neither" / single-trained-MTP-head mode, i.e. not the
gemma4 mem-shared path), and I've got most of it working but I'm stuck on the exact per-round call
order. There's no example in the repo or the vendored `llama.cpp` tree, so I'd love the intended usage.

**Setup:** `llama-cpp-2 = "0.1.154"` (metal), Qwen3.5-4B Q5_K_M, Apple M5 Max.

### What works

- `MtpSpeculative::new(target_ctx, draft_ctx, params)` with a **plain** draft context (no
  `ctx_type = MTP` — there's no public setter, and it seems unnecessary for qwen35). Constructs fine.
- `begin(prompt)`, then decoding the prompt in the target with **`logits = 1` on every position** and
  calling `process(prompt_batch)` — this satisfies the `SPC_WRN("ctx_dft pos_max=… < N-… — process()
  did not run on every prefill ubatch")` check, and afterwards `draft()` returns sensible tokens (for
  "The capital of France is" → target samples " Paris", draft head proposes ", 1").

### Where I'm stuck: the per-round order of draft / decode-in-target / process / accept

My loop per round is:

```
let drafts = spec.draft(n_past, id_last, &prompt)?;              // k candidates
// build verify batch [id_last @ n_past, d0 @ n_past+1, … d_{k-1} @ n_past+k], logits=1 on all
spec.target_context_mut().decode(&mut verify_batch)?;
spec.process(&verify_batch)?;                                     // (A)
// … compare target greedy argmax at each idx to drafts; accept longest prefix (= m) …
// roll back target KV for rejected positions
spec.accept(m)?;
```

Two things I've tried, both failing:

1. **With the in-loop `process()` at (A):** it fails on the first round with
   `process: llama_decode(ctx_dft) head=0 failed rc=-1 (pos=10)`. It looks like `draft()` has already
   advanced `ctx_dft` over the drafted positions, so `process(verify_batch)` re-decodes the same
   positions and `llama_decode` rejects it.
2. **Without the in-loop `process()`** (relying on `accept()` alone to advance the draft context): the
   **target-side** decode then fails (`decode: failed to initialize batch` / `llama_decode ret = -1`).

I can see the MTP impl has deferred-boundary cross-batch carryover
(`common/speculative.cpp` ~L403–450, `pending_pos` / `verify_g`), which is presumably why the order is
non-obvious — but I can't tell from the code alone what the caller is expected to do each round.

### Question

What's the intended per-round sequence for a full generation loop with `MtpSpeculative`? Specifically:

- Should `process()` be called **before** `draft()` (feeding the just-confirmed tokens) rather than
  after the verify decode?
- Does `draft()` already advance `ctx_dft`, so the caller must **not** decode/`process()` those same
  positions in the draft context again?
- Which batch does `process()` expect each round — the full verify batch, or only the **accepted**
  prefix (after `accept(m)`)?
- Who owns the **draft context's** KV rollback for rejected tokens — the wrapper (via `accept`), or the
  caller? (I'm rolling back the **target** context's KV myself.)

A minimal end-to-end example (prompt → N greedy tokens with acceptance) would be hugely helpful, and
I'd be happy to contribute it back as a doc/example once it works. Thanks!

**Comment by michaelburch** [comment]
Working call order, derived by getting each stage wrong first. Verified against a self-drafting Qwen3.6-27B (in-model `nextn` head) and an exported-head Gemma-4 drafter, both on Vulkan.

**Context setup**

- Target model must be loaded with `load_mtp = true`, else llama.cpp skips the `nextn` tensors and `graph_mtp` aborts on `GGML_ASSERT(layer.nextn.eh_proj)`. No setter today — #1124 adds one.
- Draft context: `with_context_type(Mtp)` + `new_context_with_ctx_other(target)`.
- Target context: `with_n_rs_seq(n_max)` if the arch has recurrent memory. See caveat 2.

**Prefill**

```
begin(prompt)
for each ubatch:
    ctx_tgt.decode(batch)
    read logits            // llama_decode returns before the GPU is done
    process(batch)
sample last row            // -> id_last
```

**Per round**

```
drafts = draft(pos, id_last, prompt)   // may legitimately return none at high p_min
batch  = [id_last@pos, drafts@pos+1..] // logits on every row
ctx_tgt.decode(batch); read logits
if drafted:
    if !mem_shared: ctx_dft.seq_rm(seq, pos, -1)   // caveat 1
    process(batch)
sample rows 0..=len(drafts), stop at first mismatch -> n_accepted
kept = pos + len(produced)
ctx_tgt.seq_rm(seq, kept, -1)
if !mem_shared: ctx_dft.seq_rm(seq, kept, -1)
if drafted && len(drafts) > 0: accept(n_accepted)
```

**Caveat 1 — the drafter may not share the target's cache.** `is_mem_shared` is `llama_get_ctx_other(ctx_dft) == ctx_tgt`, but `llama-context.cpp` keeps `ctx_other` only for `GEMMA4_ASSISTANT`, `EAGLE3` and `DFLASH` and nulls it for everything else. A head carried *inside* the model therefore always gets its own cache, which nothing trims: `draft()` writes the drafts into it, then `process()` on the verify batch starts behind it and M-RoPE rejects the batch (`X < Y` required). The caller has to trim `ctx_dft` both before `process()` and after `accept()`.

**Caveat 2 — recurrent rollback is opt-in and fails silently.** Trimming rejected drafts out of a recurrent memory (Gated DeltaNet here) needs a per-token snapshot; `llama_memory_recurrent::seq_rm` returns `false` unless `rollback <= n_rs_seq`, and `n_rs_seq` defaults to 0. `clear_kv_cache_seq` returns that flag inside its `Ok` value, so it is very easy to drop — the target then silently keeps tokens it never accepted, and the damage surfaces two rounds later as an unrelated `decode: failed to initialize batch`. Use `kv_cache_seq_rm` (which turns `false` into an `Err`) and set `n_rs_seq >= n_max`. Non-recurrent archs clamp it to 0, so it is safe to set unconditionally.

**Still unclear, and the reason this issue is worth answering in docs:**

1. `begin()` warns `ctx_dft pos_max=-1 < N-1` when called before prefill, which the sketch above does. Calling it after prefill would silence the warning but `begin()` reads as a reset. Which is intended?
2. That same warning suggests `logits=1` on every prompt position. Setting it only on the last position works and gives acceptance 0.707 at width 3; I have not measured whether per-position logits raise that.

Results, same prompt, 909-token prose prefill, 128 generated: Qwen3.6-27B Q4_K_XL 11.85 → 26.91 tok/s (2.27x), 3.07 tokens per target forward at width 3. Gemma-4-E4B Q4_0 with an exported BF16 head: 89.33 tok/s, acceptance 0.349.
